docs: add authz module tutorial (#786)
cosmos/sdk-tutorials: authz-module/index.md

---
parent:
  title: Understanding the Authz Module
order: 0
description: Use the Cosmos SDK authz module to grant authorizations from one account (the granter) to another account (the grantee).
---

# Authz Module

> The authz ("authorization") module enables a granter to grant authorizations to a grantee that allows the grantee to execute messages on behalf of the granter.

Author: Yea, that's a good idea. I've included a line explaining the difference here https://github.com/cosmos/sdk-tutorials/pull/786/commits/76f2cd0a77b9516d9ec980481b7aa00af654d65e.

## Use Authz to Grant Authorizations

By implementing the [authz module](https://docs.cosmos.network/v0.43/modules/authz/), Cosmos SDK application developers give users the ability to grant certain privileges to other users. For example, a user might want another user to vote on their behalf and so, rather than giving the other user access to their account, the user would grant an authorization that allows the other user to execute `MsgVote` on their behalf.

> How users decide to use the authz module is up to them. In one case, a validator might want to create a separate account for voting in order to keep their validator key more secure. In another case, a DAO might want to distribute authorizations to members of the DAO in which case a multisg account would grant authorizations to individual accounts and members of the DAO would be able to execute messages without requiring signatures from all the other members.

Reviewer: since we haven't talked about DAO yet, let's define it
(did you know we are working on a Cosmos SDK-approved word list so we can all use consistent spelling and style for ecosystem terminology?) see https://docs.google.com/spreadsheets/d/15raX8lfzzeeifGrDiXSFTHaEDbsMVlFJzKgnNelihjs/edit#gid=0 for https://github.com/cosmos/advocates/issues/23

> How users decide to use the authz module is up to them. In one case, a validator might want to create a separate account for voting in order to keep their validator key more secure. In another case, a DAO might want to distribute authorizations to members of the DAO in which case a multisg account would grant authorizations to individual accounts and members of the DAO would be able to execute messages without requiring signatures from all the other members.

Author: I did not know. Thanks for sharing!

> How users decide to use the authz module is up to them. In one case, a validator might want to create a separate account for voting in order to keep their validator key more secure. In another case, a DAO might want to distribute authorizations to members of the decentralized autonomous organization (DAO) in which case a multisg account would grant authorizations to individual accounts and members of the DAO would be able to execute messages without requiring signatures from all the other members.

Reviewer 2: ```suggestion
How users decide to use the authz module is up to them. In one case, a validator might want to create a separate account for voting in order to keep their validator key more secure. In another case, a decentralized autonomous organization (DAO) might want to distribute authorizations to members of the DAO in which case a multisg account would grant authorizations to individual accounts and members of the DAO would be able to execute messages without requiring signatures from all the other members.
```
Abbreviation was used before definition.

In this tutorial, you spin up a single node network using the simulation application in Cosmos SDK (`simapp`), grant an authorization to another account, and then execute a message on behalf of the granter as the grantee.

## Requirements

Before you start the tutorial, you need to install the `simd` binary.

Clone the `cosmos-sdk` repository:
```
git clone https://github.com/cosmos/cosmos-sdk
```

Change directories and check out `v0.43.0`:
```

> cd cosmos-sdk && git checkout v0.43.0

Author: All instances of `v0.43` should probably be updated to `v0.44` once `v0.44` is released and the tag is available on the cosmos sdk documentation site.

Reviewer: yeah, version maintenance is a pain. Is there a way to provide a "latest supported version" variable? :)

Author: The alternative would be using master but if changes are made to master, the tutorial would become outdated. I thought it might be best to use a specific version to prevent a potentially bad user experience.

```

Install the `simd` binary:
```
make install
```

Check to make sure the installation was successful:
```
simd version
```

You should see `0.43.0` printed to the console.

## Configuration

<!-- TODO: update tip to use `unsafe-reset-all` with better user experience -->

> ::: tip

Reviewer: thanks for using this `::: tip ` style, it's only today that I learned about it! Looks awesome (although the "i" reminds me more of "here's some information" more than a "here's a useful actionable tip")
![image](https://user-images.githubusercontent.com/10067215/131894519-5421035a-cb0e-4878-92f4-217b06f1dc40.png)

Author: You're welcome! The tip icons are included in https://github.com/cosmos/vuepress-theme-cosmos. The other options are to use "warning" and "danger".

> If you have used `simd` before, you may already have a `.simapp` directory in you home directory. If you would like to keep the previous data, either save the directory to another location or use the `--home` flag and specify a different directory for each command in the following instructions. If not, remove the previous directory (`rm -rf ~/.simapp`).

Reviewer: ```suggestion
If you have used `simd` before, you might already have a `.simapp` directory in your home directory. To keep the previous data, either save the directory to another location or use the `--home` flag and specify a different directory for each command in the following instructions. If you don't want to keep the previous data, remove the previous directory (`rm -rf ~/.simapp`).
```

Reviewer: I love words, I can talk about them all day long. Did you know that the Google developer documentation style guide is my best friend? (and is the resource that Tendermint docs rely on?)
I suggest that we use "might" to indicate possibility, see https://developers.google.com/style/word-list?hl=en#letter-m

Author: Thanks for the friendly reminder. I will re-review the style guide documentation.

Author: Thanks! I should have re-reviewed the style guide documentation.

Author: Thanks! I will try to make better use of the style guide.

:::

Run the following commands to configure the `simd` binary.

Set the chain ID:
```
simd config chain-id demo
```

Set the keyring backend:
```
simd config keyring-backend test
```

## Key Setup

> Create a couple test keys. You will use the `--recover` option so that the addresses used within the example commands below are consistent.

Reviewer: ```suggestion
Create a couple of test keys. Use the `--recover` option so that the addresses used in the following example commands are consistent.
```

Reviewer: above, below are words that we all use... but let's instead use "following" for below, see the below entry https://developers.google.com/style/word-list#letter-b

Author: Thanks! I will try to make better use of the style guide.

Add a key for Alice:
```
simd keys add alice --recover
```

Enter the following mnemonic:
```
plunge hundred health electric victory foil marine elite shiver tonight away verify vacuum giant pencil ocean nest pledge okay endless try spirit special start
```

Add a key for Bob:
```
simd keys add bob --recover
```

Enter the following mnemonic:
```
shuffle oppose diagram wire rubber apart blame entire thought firm carry swim old head police panther lyrics road must silly sting dirt hard organ
```

## Chain Setup

> The following steps will start a chain for the simulation application (`simapp`).

Reviewer: ```suggestion
The following commands start a chain for the simulation application (`simapp`).
```

Reviewer: To ensure a common voice, let's use present tense for all of our tutorials, see https://developers.google.com/style/tense 
and https://github.com/cosmos/sdk-tutorials/blob/master/CONTRIBUTING.md 

(let me know what updates we might need to the CONTRIBUTING.md file to help you and others who follow)

Author: Thanks! I think the contributing documentation is clear. I will try to make better use of the style guide next time.

Initialize the node:

> ```

Author: I think you are right. It's best practice to add a line above the code fence. Also added `bash` to each code fence.

simd init test --chain-id demo
```

Add Alice and an initial balance to the genesis file:
```

> simd add-genesis-account alice 5000000000stake

Reviewer 2: As discussed, this doesn't work without `--keyring-backend=test` :)

```

Add Bob and an initial balance to the genesis file:
```
simd add-genesis-account bob 5000000000stake
```

Generate a transaction adding Alice to the initial validator set:

> ```

Reviewer: or instead of "to add" perhaps use "that adds" 
(the -ing is so often used, in this case it sounds as if the transaction is adding Alice right now)

simd gentx alice 1000000stake --chain-id demo
```

Add the validator transaction to the genesis file:

> ```

Reviewer: ````suggestion

```bash
````

simd collect-gentxs
```

## Start Chain

You are now ready to start a single node network on your local machine.

Start the chain:
```
simd start
```

## Submit Proposal

> In order to demonstrate an authorization to vote on a governance proposal, you'll need to first create a governance proposal.

Reviewer: ```suggestion
To demonstrate an authorization to vote on a governance proposal, you must first create a governance proposal.
```

> ```

Reviewer: ````suggestion

```bash
````

Create proposal:

> ```

Reviewer: ````suggestion

```bash
````

> simd tx gov submit-proposal --title="Test Authorization" --description="Is Bob authorized to vote?" --type="Text" --deposit="10000000stake" --from alice

Reviewer 2: It's not clear to me exactly why a deposit is needed here. Maybe we should clarify? Or is this something obvious that I'm missing?

Author: Each proposal requires a deposit and `10000000stake` is the minimum deposit. Including it here prevents the need to make the deposit in a separate transaction. Once the minimum is reached, the proposal enters the voting period.

Reviewer 2: Ah yeah that makes total sense, thanks. Was just wondering if that's something we want to mention, as in, elaborating a little bit more on the parameters of this (and maybe other) command(s)?

Reviewer: as a learner for go commands (that are positional and do not require parameters before values) is to use the where method (although here it looks like these -- are flags, not parameters?) I always want to know more about flags, and the help doesn't show them (or I don't know how to make the help show all the flags available)

Author: > Was just wondering if that's something we want to mention, as in, elaborating a little bit more on the parameters of this (and maybe other) command(s)?

Sure, I can do that. This command is not directly related to the authz module but it wouldn't hurt to provide more information about it in this tutorial.

> I always want to know more about flags, and the help doesn't show them (or I don't know how to make the help show all the flags available)

You can view all flags available using `--help` (i.e. `simd tx gov submit-proposal --help`).

```

View proposal:

> ```

Reviewer: ````suggestion

```bash
````

simd q gov proposal 1
```

## Grant Authorization

> Next, the granter will need to `grant` an authorization to the grantee.
> 
> The authorization will be a "generic" authorization, meaning there are no limits on the authorization. In this case, the grantee will be able to vote as many times as they want until the granter revokes the authorization.

Reviewer: ```suggestion
Next, the granter must `grant` an authorization to the grantee.

The authorization is a "generic" authorization, meaning there are no limits on the authorization. In this case, the grantee is able to vote as many times as they want until the granter revokes the authorization.
```

> The authorization is a "generic" authorization, meaning there are no limits on the authorization. In this case, the grantee is able to vote as many times as they want until the granter revokes the authorization.

Reviewer 2: Maybe we could add a link to an overview of other authorization types besides "generic"? Not sure if that exist?

Reviewer: I wonder the same thing! IIUC, the authorization levels are full, restricted, and hidden. 
```suggestion
The authorization is a full authorization, which means there are no limits on the authorization. In this case, the grantee is able to vote as many times as they want until the granter revokes the authorization.
```

Author: The "generic" authorization is one of five supported authorization types.

```
Usage:
  simd tx authz grant <grantee> <authorization_type="send"|"generic"|"delegate"|"unbond"|"redelegate"> --from <granter> [flags]
```

I will make this more clear and point to the documentation that provides more information:
https://docs.cosmos.network/master/modules/authz/01_concepts.html#built-in-authorizations

Reviewer: today I learned about authorization types, yes, link to that section :)

Create authorization:

> ```

Reviewer: ````suggestion

```bash
````

Reviewer: @ryanchristo I won't tag suggestions for all instances, you can take it from here :)

simd tx authz grant cosmos1khljzagdncfs03x5g6rf9qp5p93z9qgc3w5dwt generic --msg-type /cosmos.gov.v1beta1.MsgVote --from alice
```

View authorization:
```
simd q authz grants cosmos1jxd2uhx0j6e59306jq3jfqs7rhs7cnhvey4lqh cosmos1khljzagdncfs03x5g6rf9qp5p93z9qgc3w5dwt /cosmos.gov.v1beta1.MsgVote
```

## Generate Transaction

> In order for the grantee to execute a message on behalf of the granter, the grantee will first need to generate an unsigned transaction where the transaction author (the `--from` address) is the granter. 

Reviewer: ```suggestion
In order for the grantee to execute a message on behalf of the granter, the grantee must first generate an unsigned transaction where the transaction author (the `--from` address) is the granter. 
```

Reviewer: in most cases, we can omit "In order for" and "In order to" and replace with "For" and "To" (just a nit for clarity)

Create unsigned transaction:
```
simd tx gov vote 1 yes --from cosmos1jxd2uhx0j6e59306jq3jfqs7rhs7cnhvey4lqh --generate-only > tx.json
```

View transaction:
```
cat tx.json
```

## Execute Transaction

> Finally, the grantee can sign and send the transaction using the `exec` command. The author of the transaction (the `--from` address) will be the grantee.

Reviewer: ```suggestion
Finally, the grantee can sign and send the transaction using the `exec` command. The author of the transaction (the `--from` address) is the grantee.
```

Sign and send transaction:
```
simd tx authz exec tx.json --from bob
```

View vote:
```
simd q gov vote 1 cosmos1jxd2uhx0j6e59306jq3jfqs7rhs7cnhvey4lqh
```

## Revoke Authorization

> If the granter would like to revoke the authorization from the grantee, the granter can use the `revoke` command.

Reviewer: ```suggestion
The granter can revoke the authorization from the grantee using the `revoke` command.
```

Revoke authorization:
```
simd tx authz revoke cosmos1khljzagdncfs03x5g6rf9qp5p93z9qgc3w5dwt /cosmos.gov.v1beta1.MsgVote --from alice
```

View authorization:
```
simd q authz grants cosmos1jxd2uhx0j6e59306jq3jfqs7rhs7cnhvey4lqh cosmos1khljzagdncfs03x5g6rf9qp5p93z9qgc3w5dwt /cosmos.gov.v1beta1.MsgVote
```

## 🎉 Congratulations 🎉

> By completing this tutorial, you have learned how to use the authz module.

Reviewer: ```suggestion
By completing this tutorial, you have learned how to use the authz module:

- Configured and used the simulation application (simapp)
- Created a government proposal
- Created an authorization 
- Generated and executed an unsigned transaction
- Revoked authorization

```

> To learn more about the authorization module and different types of authorizations, check out the [authz module documentation](https://docs.cosmos.network/v0.43/modules/authz/).

Reviewer 2: Link didn't work for me, not sure if it's already up?

Reviewer: what is our version URL best practice? do we want to use master so the link automatically works? or we use the current version and then auto promote the topic for next version?

Author: It looks like there might be an issue with the deployment at the moment. :grimacing: This is the correct link. Also, I thought it might be best to use a specific version to prevent a potentially bad user experience if changes are made to master.

```

Add Bob and an initial balance to the genesis file:

```bash
simd add-genesis-account bob 5000000000stake --keyring-backend test
```

Generate a transaction to add Alice to the initial validator set:

```bash
simd gentx alice 1000000stake --chain-id demo
```

Add the validator transaction to the genesis file:

```bash
simd collect-gentxs
```

## Start Chain

You are now ready to start a single node network on your local machine.

Start the chain:

```bash
simd start
```

## Submit Proposal

To demonstrate an authorization to vote on a governance proposal, you must first create a governance proposal. The following command creates a text proposal that includes the minimun deposit, which allows the governance proposal to immediately enter the voting period. For more information about the command and the flag options, run `simd tx gov submit-proposal --help`.

Create proposal:

```bash
simd tx gov submit-proposal --title="Test Authorization" --description="Is Bob authorized to vote?" --type="Text" --deposit="10000000stake" --from alice
```

View proposal:

```bash
simd q gov proposal 1
```

## Grant Authorization

Next, the granter must `grant` an authorization to the grantee.

> The authorization is a "generic" authorization, which means there are no limits on the authorization. Other [authorization types](https://docs.cosmos.network/v0.44/modules/authz/01_concepts.html#built-in-authorizations) include "send", "delegate", "unbond", and "redelegate" in which case a limit on the amount of tokens can be set by the granter. In this case, the grantee is able to vote as many times as they want until the granter revokes the authorization.

Reviewer 2: Small suggestion for clarity.

Also, just wondering, but "delegate", "unbond" and "redelegate" aren't listed in the built-in authorizations overview. Does [that doc](https://github.com/cosmos/cosmos-sdk/blob/master/x/authz/spec/01_concepts.md#Built-in-Authorizations) need to be updated?

```suggestion
The authorization is a "generic" authorization, which is one that takes a message type (such as `MsgVote`) as a parameter and allows the grantee unlimited authorization to execute that message on behalf of the granter. Other [authorization types](https://docs.cosmos.network/v0.44/modules/authz/01_concepts.html#built-in-authorizations) include "send", "delegate", "unbond", and "redelegate" in which case a limit on the amount of tokens can be set by the granter. In this case, the grantee is able to vote as many times as they want until the granter revokes the authorization.
```

Author: > Also, just wondering, but "delegate", "unbond" and "redelegate" aren't listed in the built-in authorizations overview. Does that doc need to be updated?

Yup, looks like `StakeAuthorization` needs to be added. `StakeAuthorization` defines authorization for delegate, unbond, and redelegate, which are listed out here as the options presented when using the CLI.
